### ISSUE TITLE
running ${LIBEXEC_DIR}/daps-xmlwellformed --xinclude ${MAIN} fails

### ISSUE THREAD

**Issue by lopippo** [tomschr/daps#687]
### Problem description
Greetings, 

now that we have solved the libxslt-related issue (see #676), I started working on the Debian package for daps. I used my own software documentation and found that I could not run the usual commands any more, I would get a "Fatal error:" without explanation.

After some printf-debugging of the bin/daps script, I found that line (1740, I think):

```
CHECK_WELLFORMED=$(PYTHONWARNINGS="ignore" ${LIBEXEC_DIR}/daps-xmlwellformed --xinclude ${MAIN} 2>&1
```

was involved. Removing `--xinclude` from the line above would solve the problem.

Comment, the error reporting code below that line is of no use, since it does not report any error, actually:

```
daps --force -d /home/rusconi/devel/minexpert2/development/doc/user-manual/DC-user-manual --verbosity=3 pdf
Fatal error:
```

I'll patch daps, as this error does not seem of major importance, though.

Thank you for listening :-)

Sincerely,
Filippo

**Comment by fsundermeyer** [comment]
Hi,

thank you for the report.
Removing `--xinclude` from the daps-xmlwellformed script is not a good idea ;-).

Let me explain why I introduced this check:
All operations on XML files in DAPS require parsing via XSLT. This already starts with some basic operations like obtaining the language from `xml:lang` in the doc's root element or the root element itself. These operations only work if the XML is well-formed (it does not ncessarily need to validate at this stage, but well-formedness is a must). Since DAPS is mostly based on bash and GNU make, error handling sometimes is a challenge (as you also have experienced above). So rather than having XSLT calls fail silently or with too little information to be helpful, the first thing DAPS does is to check for well-formedness. In case of errors users get proper error messages exactly describing the error, so it is easy to fix.

However, this requires to check _all_ files a document consists of. With DocBook, you usually have a master file defining the structure and including more files via xi:include (see https://github.com/openSUSE/daps/blob/main/doc/xml/MAIN.DAPS.xml for an example).
If you leave out `--xinclude` the script would only check the single master file and ignore all other files.

Would you be so kind to run the following command directly from the DAPS checkout directory and report the error message?

`libexec/daps-xmlwellformed --xinclude doc/xml/MAIN.DAPS.xml`

As for the second (silent) error--does this also occur when building html?
`daps --force -d /home/rusconi/devel/minexpert2/development/doc/user-manual/DC-user-manual --verbosity=3 html`

Usually these silent fatal errors are caused by FOP because it does not like an SVG. Please run daps with `--debug` rather than with verbosity 3--this will also enable JAVA error messages.

Frank

**Comment by lopippo** [comment]
Greetings, Frank,
thank you for your quick response and for the explanation.

Checked out version 3.3.1 as a git clone operation.

```
% libexec/daps-xmlwellformed --xinclude doc/xml/MAIN.DAPS.xml
No output, 
% echo $? gives 0.
```

To test the installed setup, I ./configure --prefix=/tmp/usr and then make install.
Now:

```
% export PATH=/tmp/usr/bin:${PATH} 
```

and I can test the daps program.

```
% cd <daps_top_source_dir>
/tmp/usr/share/daps/libexec/daps-xmlwellformed --xinclude doc/xml/MAIN.DAPS.xml
No output

% cd <minexpert2_top_source_dir>
/tmp/usr/share/daps/libexec/daps-xmlwellformed --xinclude doc/user-manual/xml/MAIN-user-manual.xml
No output
```

Back to `<daps_top_source_dir>`:

```
% daps -d doc/DC-daps-tutorial pdf
PDF book built with REMARKS=0, DRAFT=no and META=0:
/home/rusconi/tmp/daps/doc/build/daps-tutorial/art.tutorial_color_en.pdf
Works!
```

Now to `<minexpert2_top_source_dir>`:

```
% daps -d doc/user-manual/DC-user-manual pdf 
Fatal error: <no error message>
```

So that seems to pinpoint something wrong with my DC-user-manual file.

This file's contents:

```
## MAIN file of the document, mandatory
## The MAIN file is the central DocBook XML file of your document that
## references all other DocBook XML files that may be part of your document.
## By convention, its name starts with "MAIN.".
MAIN="MAIN-user-manual.xml"
ROOTID=""
STYLEROOT="xslt"
FOP_CMD_OPTIONS="-c /home/rusconi/devel/minexpert2/development/doc/user-manual/fop.xconf"
```

I tried the same command after additively commenting out the lines from the bottom, always getting a failure.
So, the last file's contents I tried was this single line:

    MAIN="MAIN-user-manual.xml"

and I still got the error. So that seems to pinpoint a problem in "MAIN-user-manual.xml".

But, astonishingly, the following command does not fail:
/tmp/usr/share/daps/libexec/daps-xmlwellformed --xinclude doc/user-manual/xml/MAIN-user-manual.xml

So, I tried one variant of the daps command line:

```
daps -m doc/user-manual/xml/MAIN-user-manual.xml pdf
Fatal error: <no error message>
```

That MAIN-user-manual.xml file's contents:

```
<?xml-model href="http://docbook.org/xml/5.0CR3/rng/docbookxi.rng"
schematypens="http://relaxng.org/ns/structure/1.0"?>

<!--
<?xml-model href="http://docbook.org/xml/5.1CR3/rng/docbook.rng"
schematypens="http://purl.oclc.org/dsdl/schematron"?>
-->

<book xmlns="http://docbook.org/ns/docbook" version="5.0"
xmlns:xlink="http://www.w3.org/1999/xlink" 
xmlns:xi="http://www.w3.org/2001/XInclude" 
xml:id="book.template" 
xml:lang="en">

<xi:include href="book-info.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<dedication>

<para>Dedication text</para>

<para>To all the readers who helped me with this manual.</para>

</dedication>

<xi:include href="preface.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="generalities.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="graphical-user-interface.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="mass-data-integrations.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="mass-spectral-deconvolutions.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="isotopic-cluster-computations.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="processing-flows.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="record-data-mining-discoveries.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="gpl-3.0.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

<xi:include href="colophon.xml" xmlns:xi="http://www.w3.org/2001/XInclude">
</xi:include>

</book>
```

As for the --debug flag on the daps command line, that does not add any output to the Fatal error: line.

And now I am at a loss...
Any idea ?
Sincerely,
Filippo

**Comment by lopippo** [comment]
One more idea: I changed directory to the xml directory containing all the xinclude'd xml files and ran

    % for file in *.xml; do /tmp/usr/share/daps/libexec/daps-xmlwellformed --xinclude ${file}; done

Not a single error.

Sincerely,
Filippo

**Comment by lopippo** [comment]
Gone forward with testing (which reminds me debugging LaTeX code years ago :-)

I have edited the MAIN-user-manual.xml file and noticed that I had to remove all the xi:include elements for the pdf file to be generated (although only with the dedication text). So clearly there was a problem with these statements. What is puzzling is that all the xi:include'd files were not at fault with daps-xmlwellformed (previous post).

Sincerely,
Filippo

**Comment by tomschr** [comment]
Hi Filippo,

just a wild guess: when you build your package in Debian, maybe your FOP version is a bit too old? I use FOP version 2.7.

Another idea: If it would be possible, could you share your documentation as an archive? I could have a look and maybe it's something different.

**Comment by tomschr** [comment]
Just another wild guess. I saw this line in your DC file:

    FOP_CMD_OPTIONS="-c /home/rusconi/devel/minexpert2/development/doc/user-manual/fop.xconf"

I think this line is a bit suspicious and may be in conflict with the same option in `libexec/daps-fop`, our shell wrapper. When calling daps, it uses that shell wrapper for FOP and not the original `/usr/bin/fop` script. 

Could you replace the above line all together and insert this:

    FOP_CONFIG_FILE="/home/rusconi/devel/minexpert2/development/doc/user-manual/fop.xconf"

Does it change anything?

**Comment by lopippo** [comment]
> Just another wild guess. I saw this line in your DC file:
> 
> ```
> FOP_CMD_OPTIONS="-c /home/rusconi/devel/minexpert2/development/doc/user-manual/fop.xconf"
> ```
> 
> I think this line is a bit suspicious and may be in conflict with the same option in `libexec/daps-fop`, our shell wrapper. When calling daps, it uses that shell wrapper for FOP and not the original `/usr/bin/fop` script.
> 
> Could you replace the above line all together and insert this:
> 
> ```
> FOP_CONFIG_FILE="/home/rusconi/devel/minexpert2/development/doc/user-manual/fop.xconf"
> ```
> 
> Does it change anything?

No, it gives exactly the same errror. I'll provide you with an archive of the documenation project (big file).
Thank you !
Filippo

**Comment by lopippo** [comment]
> Hi Filippo,
> 
> just a wild guess: when you build your package in Debian, maybe your FOP version is a bit too old? I use FOP version 2.7.
> 
> Another idea: If it would be possible, could you share your documentation as an archive? I could have a look and maybe it's something different.

My fop version is 2.6.

Filippo

**Comment by lopippo** [comment]
Greetings,
here is a link to the archive https://filesender.renater.fr/?s=download&token=ebb54482-fbfc-4b51-b889-e23dffc922e2 or this one
https://filesender.renater.fr/download.php?token=ebb54482-fbfc-4b51-b889-e23dffc922e2&files_ids=18547459
to the archive.

Build works only if I remove the --xinclude flag from the command line inside /usr/bin/daps.

Thank you so much for trying to help me with this :-)

Sincerely,
Filippo

**Comment by tomschr** [comment]
Thanks Filippo for the link! That has helped a lot. :+1: 

Actually nothing related to daps itself (or maybe indirectly). However, I see the following issues in your files:

* You seem to refer to DocBook 5.0CR3 (file `MAIN-user-manual.xml` inside the processing instruction). Don't do that, it's an old version. You should use at least version 5.1, or better 5.2. However, version 5.2 is probably not available in Debian. This alone does not cause issues. It's more of a hint for your (XML) editor.
* Add the following line to your DC file to force validation with DocBook 5.1:

        DOCBOOK5_RNG_URI="http://docbook.org/xml/5.1/rng/docbookxi.rng"

  Or use even version 5.2, if available.

* The file `xml/javascript-reference.txt` doesn't exist.
* Your structure uses text files which use external entities, for example in file `scripting-reference.xml`, entity `jsreferenceent`. This method is outdated. Nowadays, to include text files we use XIncludes, for example `<xi:include href="javascript-reference.txt" parse="text"/>`
* The file `book-info.xml` contains `<info>` as root element. This is invalid in DocBook. Better add them in your main file.

I would also recommend to change the following issues. They are not critical, but can be cumbersome or error-prone:

* Your XInclude namespace (`xmlns:xi="http://www.w3.org/2001/XInclude"`) needs to be referred only once in the root element. Once defined in the root element, the namespace (`xi` in our example) is propagated through the entire tree. As such, you only need to write `<xi:include href="..."/>` and you are done. No need to repeat yourself. :wink: 
* Define a *central entity file* which are included in every XML file. That way you define them only once and they are available everywhere.
* You should move the `<!ENTITY % sgml.features "IGNORE">` and `<!ENTITY % xml.features "INCLUDE">` in your central entity file.

I haven't gone through all of your files, but that's the first things that I recognized. :slightly_smiling_face:

**Comment by lopippo** [comment]
Greetings, Tom,

I am grateful to your for your perusal of my code and for the suggestions. I'll implement these right now. However, there seems to be a showstopper here: in testing, the latest docbook5-xml version is 5.0.

I'll try to see if I can succeed in updating it or work together with other Debianites of the XML/SGML Group.
I'll see if your suggestions fix the problem at hand.

Sincerely,
Filippo

**Comment by tomschr** [comment]
> in testing, the latest docbook5-xml version is 5.0.

You can use that version for the time being. It might be easier. My examples used 5.1, just because it's a bit more modern.

For your documents, I suppose 5.0 would be sufficient.

> I'll see if your suggestions fix the problem at hand.

Probably I missed some. Tomorrow I can look into it again.

**Comment by tomschr** [comment]
I'm sorry if it takes some efforts. I know it looks overwhelming at first, but I'm sure you end up with a more robust documentation. Keep up the good work!

There is one more issue:

* In your entity files, you use DocBook elements. That's fine, but you haven't included the DocBook5 namespace in the root element. See an example in our [`generic-entities.ent` file](https://github.com/SUSE/doc-sle/blob/main/xml/generic-entities.ent#L44).

A tip how you could fix all the errors:

1. Create a central entity file (for example, you could name it `entities.ent`) and add all other external entities.
    If you want to have one file which contains all of your entities or split them up into different files is up to you. Just a matter of taste. :wink: 
    However, the idea is to include this central entity file in _all_ of your XML files in the DOCTYPE declaration.
1. In your main file, add the following DOCTYPE:

        <!DOCTYPE <YOUR_ROOT_ELEMENT> [
           <!ENTITY % entities SYSTEM "entities.ent">
           %entities;
        ]>

   Add these lines to _every_ XML file.

1. Insert all content of `book-info.xml` into your main file. Remove the file `book-info.xml`, you don't need it anymore.
1. In your main file, comment all XInclude elements.
1. Validate your document with daps. It should be valid now. If not, look into your entities (watch for missing DocBook5 namespace in elements).
1. Enable the first XInclude element again and validate.
   a. Does it valid? Great! Enable the next XInclude element and try it again.
   b. Does you get errors? Now you know it has something to do with the file that you just xincluded. You need to fix that.

   The idea here is to enable each XInclude step by step. You may get missing IDs or other errors. That's fine. As long as you don't get any "fatal error" just continue.
1. Proceed with the next XInclude element until you have reached the last.

With that procedure you should find and eliminate all of your errors in the files.

**Comment by lopippo** [comment]
Greetings, Tom,

> I'm sorry if it takes some efforts.

Oh no! you are the one doing efforts for the sake of beauty and elegance and technical excellence! Thank you! I'll look into your comments soon, good part of today is microbiology research wet bench work :-)
Most sincerely,
Filippo

**Comment by tomschr** [comment]
By the way, one more thing regarding building/packaging on Debian... this repo contains a `debian` folder which some files related to packaging. It could be helpful to package daps for Debian (we take pull requests! :wink: )

I don't have much experience with Debian, but it seems, some packages for XML work and their dependencies seem to bit outdated. We took some effort for openSUSE to package or update them. If you really want to do the same for Debian, it can take some work. 

Another alternative would be to use [daps2docker](https://github.com/openSUSE/daps2docker) on Debian. It's a Docker container, based on openSUSE, which you can also run on Debian.

**Comment by lopippo** [comment]
> By the way, one more thing regarding building/packaging on Debian... this repo contains a `debian` folder which some files related to packaging. It could be helpful to package daps for Debian (we take pull requests! wink )

Well, this is what I am set to do: package daps for Debian. I will clone this daps instance, update all the debian/* files in it. I was using my own documentation to test the package. I'll let you know.
Sincerely,
Filippo

**Comment by lopippo** [comment]
Greetings, Tom,

Well, your advice proved exactly to the point. By implementing all your modifications, and narrowing down error by error, I could make my documentation. I would say that most of the problems were related to the missing  xmlns statement in the various ENTITY_s I have defined. There was also one missing label which was referenced.

Now, I see one suggestion for an improvement of daps: clearly, the error "Fatal error:" that is returned is only useful when there is a problem of "DTD validation", that is, an element not in the right place or garbled start and end of element. But for a large number of the errors I got, there  was no error detail whatsoever. That was hard, because it is nice to have a hint at what code line or code paragraph to look at to fix the problem. 

Thank your for your support!

Feel free to close the issue.

Sincerely,
Filippo

**Comment by tomschr** [comment]
Hi Filippo,

> Well, this is what I am set to do: package daps for Debian. I will clone this daps instance, update all the debian/* files in it. I was using my own documentation to test the package. I'll let you know.

Great! The folder might be a bit outdated, but it could be a good start.

> one suggestion for an improvement of daps clearly, the error "Fatal error:" that is returned is only useful when there is a problem of "DTD validation"

Yes, that's unfortunate. It seems the culprit is the `daps-xmlwellformed` script. It uses the lxml library. I had always issues with it as the lxml library has some limitation related to reporting/logging.

Daps is build around some XML tools and sometimes they have some limitations (and also sometimes bugs). It would be helpful if we could reduce it to a small test case which caused this fatal error.

> But for a large number of the errors I got, there was no error detail whatsoever.

Yes, that's also unfortunate. Normally this should be reported by daps. If you want a quick test, you could run `xmllint` against your XML files for the time being:

    xmllint --noent --noout  xml/*.xml

That will report some problems in your files (they are already covered above).

@fsundermeyer I'm not sure if we should close this bug for the time being or leave it open.

**Comment by fsundermeyer** [comment]
Commit 60671af106d should fix this problem. The error described above is no "wellformdness" error, so we can safely ignore it at this point. The real validation check (that is a prerequisite for every socumentation build) will properly catch the error and will produce a helpful error messages  (tested with the content from https://github.com/openSUSE/daps/issues/687#issuecomment-1281180645)
This fix will be published with [DAPS 3.3.2](https://github.com/openSUSE/daps/tree/3.3.2-release) which is already in the works.

**Comment by lopippo** [comment]
Greetings, Frank,

> This fix will be published with [DAPS 3.3.2](https://github.com/openSUSE/daps/tree/3.3.2-release) which is already in the works.

Great, thank you!
Sincerely
Filippo
